openvdb_points/tools/PointCount.h

```diff
@@ -0,0 +1,223 @@
+
+#include <openvdb/openvdb.h>
+
+#include <openvdb_points/tools/AttributeSet.h>
+#include <openvdb_points/tools/PointDataGrid.h>
+#include <openvdb_points/tools/PointAttribute.h>
+
+#include <boost/ptr_container/ptr_vector.hpp>
+
+#include <tbb/parallel_reduce.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+template <typename IterT>
+inline Index64 iterCount(const IterT& iter)
+{
+    Index64 size = 0;
+    for (IterT newIter(iter); newIter; ++newIter, ++size) { }
+    return size;
+}
+
+
+template <typename PointDataTreeT>
+Index64 pointCount(const PointDataTreeT& tree)
+{
+    Index64 size = 0;
+    for (typename PointDataTreeT::LeafCIter iter = tree.cbeginLeaf(); iter; ++iter) {
+        size += iter->pointCount();
+    }
+    return size;
+}
+
+
+template <typename PointDataTreeT>
+Index64 activePointCount(const PointDataTreeT& tree)
+{
+    Index64 size = 0;
+    for (typename PointDataTreeT::LeafCIter iter = tree.cbeginLeaf(); iter; ++iter) {
+        size += iter->onPointCount();
+    }
+    return size;
+}
+
+
+template <typename PointDataTreeT>
+Index64 inactivePointCount(const PointDataTreeT& tree)
+{
+    Index64 size = 0;
+    for (typename PointDataTreeT::LeafCIter iter = tree.cbeginLeaf(); iter; ++iter) {
+        size += iter->offPointCount();
+    }
+    return size;
+}
+
+
+namespace point_count_internal {
+
+template <  typename PointDataTreeT,
+            typename ValueIterT,
+            typename FilterFromLeafT>
+struct PointCountOp
+{
+    typedef typename tree::LeafManager<PointDataTreeT>      LeafManagerT;
+    typedef typename PointDataTreeT::LeafNodeType           LeafT;
+    typedef IndexIterTraits<PointDataTreeT, ValueIterT>     IndexIteratorFromLeafT;
+    typedef typename IndexIteratorFromLeafT::Iterator       IndexIterator;
+    typedef typename FilterFromLeafT::Filter                Filter;
+    typedef FilterIndexIter<IndexIterator, Filter>          Iterator;
+
+    PointCountOp(const PointDataTreeT& tree,
+                 const LeafManagerT& leafs,
+                 const FilterFromLeafT& filterFromLeaf)
+        : mTree(tree)
+        , mLeafs(leafs)
+        , mFilterFromLeaf(filterFromLeaf) { }
+
+    Index64 operator()(const tbb::blocked_range<size_t>& range, Index64 size) const {
+
+        for (size_t n = range.begin(); n != range.end(); ++n) {
+            const LeafT& leaf = mLeafs.leaf(n);
+
+            IndexIterator indexIterator(IndexIteratorFromLeafT::begin(leaf));
+            Filter filter(mFilterFromLeaf.fromLeaf(leaf));
+            Iterator iter(indexIterator, filter);
+            size += iterCount<Iterator>(iter);
+        }
+
+        return size;
+    }
+
+    static Index64 join(Index64 size1, Index64 size2) {
+        return size1 + size2;
+    }
+
+private:
+    const PointDataTreeT& mTree;
+    const LeafManagerT& mLeafs;
+    const FilterFromLeafT& mFilterFromLeaf;
+}; // struct PointCountOp
```

Comment:
I think this Op can be a lot simpler and remove the need to const cast the point data tree in threadedFilterPointCount:

    PointCountOp(const FilterFromLeafT& filterFromLeaf)
        : mFilterFromLeaf(filterFromLeaf) {}

    Index64 operator()(const typename LeafManagerT::LeafRange& range, Index64 size) const
    {
        for (typename LeafManagerT::LeafRange::Iterator leaf = range.begin(); leaf; ++leaf)
        {
            IndexIterator indexIterator(IndexIteratorFromLeafT::begin(*leaf));
            Filter filter(mFilterFromLeaf.fromLeaf(*leaf));
            Iterator iter(indexIterator, filter);
            size += iterCount<Iterator>(iter);
        }

        return size;
    }

    static Index64 join(Index64 size1, Index64 size2) {
        return size1 + size2;
    }
    private:
        const FilterFromLeafT& mFilterFromLeaf;

and called like so:

    typedef point_count_internal::PointCountOp2< PointDataTreeT, ValueIterT, FilterFromLeafT> PointCountOp;

    typename tree::LeafManager<const PointDataTreeT> leafManager(tree);
    const PointCountOp pointCountOp(filterFromLeaf);
    return tbb::parallel_reduce(leafManager.leafRange(), Index64(0), pointCountOp,  PointCountOp::join);